src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -45,42 +45,86 @@
-    if (!check) {
-        ids_[tx] = i;
-        dis_[tx] = d;
-    } else {
-        ids_[tx] = -1;
-        dis_[tx] = -1.0f;
+    __shared__ int64_t ids_before_s[1024];
+    __shared__ int32_t ids_block_s[1024];
+
+    ids_before_s[threadIdx.x] = ids_before[tx];
+    int64_t i = ids_before_s[threadIdx.x];
+    ids_block_s[threadIdx.x] = 1 - (bs[i >> 3] && (0x1 << (i & 0x7)));
+    ids_block[tx] = ids_block_s[threadIdx.x];
+}
+
+__global__ void
+prescan(int32_t* g_odata, int32_t* g_idata, int n) {
+    extern __shared__ int32_t temp[];
+    int tx = blockIdx.x * blockDim.x + threadIdx.x;
+    int ty = blockIdx.y * blockDim.y + threadIdx.y;
+    int offset = 1;
+    temp[2 * tx] = g_idata[2 * tx + ty * n];
+    temp[2 * tx + 1] = g_idata[2 * tx + 1 + ty * n];
+    for (int d = n >> 1; d > 0; d >>= 1) {
+        __syncthreads();
+        if (tx < d) {
+            int ai = offset * (2 * tx + 1) - 1;
+            int bi = offset * (2 * tx + 2) - 1;
+            temp[bi] += temp[ai];
+        }
+        offset *= 2;
     }
-    __syncthreads();
     if (tx == 0) {
-        int j = 0, k = 0;
-        while (j < k1 && k < k2) {
-            while (ids_[k] == -1) k++;
-            if (k >= k2)
-                break;
-            ids[ty * k1 + j] = ids_[k];
-            dis[ty * k1 + j] = dis_[k];
-            j++;
-            k++;
-        }
-        if (j != k1) {
-            ids_before[0] = -1;  // destroy answer
+        temp[n - 1] = 0;
+    }
+    for (int d = 1; d < n; d *= 2) {
+        offset >>= 1;
+        __syncthreads();
+        if (tx < d) {
+            int ai = offset * (2 * tx + 1) - 1;
+            int bi = offset * (2 * tx + 2) - 1;
+            int32_t t = temp[ai];
+            temp[ai] = temp[bi];
+            temp[bi] += t;
         }
     }
     __syncthreads();
+    g_odata[2 * tx + ty * n] += temp[2 * tx];
+    g_odata[2 * tx + 1 + ty * n] += temp[2 * tx + 1];
+}
+
+__global__ void
+write_back(const int32_t* pre_ids_block, int64_t* ids_before, float* dis_before, int64_t* ids_after, float* dis_after,
+           int k1, int k2) {
+    int tx = blockIdx.x * blockDim.x + threadIdx.x;
+    int ty = blockIdx.y * blockDim.y + threadIdx.y;
+
+    extern __shared__ int32_t temp[];
+    for (int i = tx; i < k2; i += k1) {
+        temp[i] = pre_ids_block[ty * k2 + i];
+    }
+    __syncthreads();
+    int left = 0;
+    int right = k2;
+    while (left < right) {
+        int mid = (left + right) / 2;
+        if (temp[mid] == tx + 1) {
+            right = mid;
+        } else if (temp[mid] > tx + 1) {
+            right = mid;
+        } else if (temp[mid] < tx + 1) {
+            left = mid + 1;
+        }
+    }
+    if (left < k2) {
+        ids_after[ty * k1 + tx] = ids_before[ty * k2 + left];
+        dis_after[ty * k1 + tx] = dis_before[ty * k2 + left];
+    } else {
+        ids_after[ty * k1 + tx] = -1;
+        dis_after[ty * k1 + tx] = 1.0f / 0.0f;
+        ids_before[0] = -1;  // destory old ans;
```

Comment:
In my opinion, a better way is just scan k2 data and write back into k1 result. This way use k1 result to find data from k2 data, which may cause bad performance when bitset is dense.

Reply:
Good Advise.